File: packages/skillet-eval-python/src/skill_eval/config/registry.py
from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Literal

ProviderName = Literal["anthropic", "openai", "google"]


@dataclass
class ModelEntry:
    id: str
    label: str
    tag: Literal["balanced", "reasoning", "fast", "code"]


@dataclass
class ProviderRegistryEntry:
    label: str
    env_keys: list[str]
    default_model: str
    models: list[ModelEntry] = field(default_factory=list)


PROVIDER_REGISTRY: dict[str, ProviderRegistryEntry] = {
    "anthropic": ProviderRegistryEntry(
        label="Anthropic",
        env_keys=["ANTHROPIC_API_KEY"],
        default_model="claude-sonnet-4-6",
        models=[
            ModelEntry("claude-sonnet-4-6", "Sonnet 4.6", "balanced"),
            ModelEntry("claude-opus-4-6", "Opus 4.6", "reasoning"),
            ModelEntry("claude-haiku-4-5", "Haiku 4.5", "fast"),
        ],
    ),
    "openai": ProviderRegistryEntry(
        label="OpenAI",
        env_keys=["OPENAI_API_KEY"],
        default_model="gpt-5.4",
        models=[
            ModelEntry("gpt-5.4", "GPT-5.4", "balanced"),
            ModelEntry("o4-mini", "o4-mini", "reasoning"),
            ModelEntry("gpt-5.3-codex", "GPT-5.3 Codex", "code"),
        ],
    ),
    "google": ProviderRegistryEntry(
        label="Google",
        env_keys=["GOOGLE_API_KEY", "GEMINI_API_KEY"],
        default_model="gemini-3.1-pro-preview",
        models=[
            ModelEntry("gemini-3.1-pro-preview", "Gemini 3.1 Pro", "balanced"),
            ModelEntry("gemini-3-flash-preview", "Gemini 3 Flash", "fast"),
            ModelEntry("gemini-3.1-flash-lite-preview", "Gemini 3.1 Flash Lite", "fast"),
        ],
    ),
}

DEFAULT_MODELS: dict[str, str] = {
    k: v.default_model for k, v in PROVIDER_REGISTRY.items()
}

ENV_KEY_MAP: dict[str, list[str]] = {
    k: v.env_keys for k, v in PROVIDER_REGISTRY.items()
}

MODEL_PREFIXES: list[tuple[re.Pattern[str], ProviderName]] = [
    (re.compile(r"^claude-"), "anthropic"),
    (re.compile(r"^gpt-"), "openai"),
    (re.compile(r"^o\d"), "openai"),
    (re.compile(r"^chatgpt-"), "openai"),
    (re.compile(r"^gemini-"), "google"),
]


def infer_provider(model: str) -> ProviderName:
    for pattern, provider in MODEL_PREFIXES:
        if pattern.search(model):
            return provider
    raise ValueError(
        f'Cannot infer provider for model "{model}". '
        f"Use provider:model syntax (e.g. openai:{model}) or add it to the config file."
    )

File: packages/skillet/src/eval/config/registry.ts
export interface ModelEntry {
	id: string;
	label: string;
	tag: "balanced" | "reasoning" | "fast" | "code";
}

export interface ProviderRegistryEntry {
	label: string;
	envKeys: string[];
	defaultModel: string;
	models: ModelEntry[];
}

export const PROVIDER_REGISTRY: Record<string, ProviderRegistryEntry> = {
	anthropic: {
		label: "Anthropic",
		envKeys: ["ANTHROPIC_API_KEY"],
		defaultModel: "claude-sonnet-4-6",
		models: [
			{ id: "claude-sonnet-4-6", label: "Sonnet 4.6", tag: "balanced" },
			{ id: "claude-opus-4-6", label: "Opus 4.6", tag: "reasoning" },
			{ id: "claude-haiku-4-5", label: "Haiku 4.5", tag: "fast" },
		],
	},
	openai: {
		label: "OpenAI",
		envKeys: ["OPENAI_API_KEY"],
		defaultModel: "gpt-5.4",
		models: [
			{ id: "gpt-5.4", label: "GPT-5.4", tag: "balanced" },
			{ id: "o4-mini", label: "o4-mini", tag: "reasoning" },
			{ id: "gpt-5.3-codex", label: "GPT-5.3 Codex", tag: "code" },
		],
	},
	google: {
		label: "Google",
		envKeys: ["GOOGLE_API_KEY", "GEMINI_API_KEY"],
		defaultModel: "gemini-3.1-pro-preview",
		models: [
			{ id: "gemini-3.1-pro-preview", label: "Gemini 3.1 Pro", tag: "balanced" },
			{ id: "gemini-3-flash-preview", label: "Gemini 3 Flash", tag: "fast" },
			{ id: "gemini-3.1-flash-lite-preview", label: "Gemini 3.1 Flash Lite", tag: "fast" },
		],
	},
};

export const DEFAULT_MODELS: Record<string, string> = Object.fromEntries(
	Object.entries(PROVIDER_REGISTRY).map(([k, v]) => [k, v.defaultModel]),
);

export const ENV_KEY_MAP: Record<string, string[]> = Object.fromEntries(
	Object.entries(PROVIDER_REGISTRY).map(([k, v]) => [k, v.envKeys]),
);

type ProviderName = "anthropic" | "openai" | "google";

const MODEL_PREFIXES: [RegExp, ProviderName][] = [
	[/^claude-/, "anthropic"],
	[/^gpt-/, "openai"],
	[/^o\d/, "openai"],
	[/^chatgpt-/, "openai"],
	[/^gemini-/, "google"],
];

/** Infer the provider from a model name, or throw if unrecognized. */
export function inferProvider(model: string): ProviderName {
	for (const [pattern, provider] of MODEL_PREFIXES) {
		if (pattern.test(model)) return provider;
	}
	throw new Error(
		`Cannot infer provider for model "${model}". ` +
			`Use provider:model syntax (e.g. openai:${model}) or add it to the config file.`,
	);
}
